### DISCUSSION TITLE
What is included in the hash when `CCACHE_COMPILERCHECK` is set to `content`?

### DISCUSSION THREAD

**Discussion by felixsinger** [ccache/ccache#1311] [Questions & answers] [answered]
Hi there,

we experienced long build times with Clang when we set `CCACHE_COMPILERCHECK` to `content`. Tracking down the issue, the issue relies in how we build our Clang toolchain.

Building Clang with `LLVM_LINK_LLVM_DYLIB` disabled, which is the default configuration, the size of the Clang binary (in our case `/bin/clang-16`) results in around 155 MB, which explains the long build times as ccache hashes the compiler binary. Enabling this option, splits the Clang binary into a CLI tool with 195 KB in size and separate shared libraries, which decreases the build time (rather hashing times) significantly. For details, see here https://review.coreboot.org/c/coreboot/+/76401.

However, when `LLVM_LINK_LLVM_DYLIB` is enabled, this also results in that our own vendor string with the toolchain version is not compiled in the Clang binary anymore, which I guess makes it hard for ccache to differentiate between various toolchain versions. I checked this by running `strings clang-16 | grep coreboot`, which doesn't give any results.

So i am wondering if ccache just hashes this Clang binary when it's running in the `content` mode or if it includes any other information.

Thank you in advance!

**Comment by jrosdahl** [comment, marked as answer]
> So i am wondering if ccache just hashes this Clang binary when it's running in the `content` mode or if it includes any other information.

Yes, with `CCACHE_COMPILERCHECK= content` ccache hashes only the content of the compiler binary, not any of its dependencies. To pick up your vendor string in `LLVM_LINK_LLVM_DYLIB` mode you would currently have to use either something like `CCACHE_COMPILERCHECK="%compiler% -v"` or `CCACHE_COMPILERCHECK="string:some string you compute yourself"`.

**Reply by felixsinger** [reply]
Alright, thank you. I will consider doing that and I will reach out to the LLVM project and ask if they could add the vendor string to the Clang binary.
